designcompose/src/main/java/com/android/designcompose/squoosh/SquooshAnimate.kt

```diff
@@ -0,0 +1,395 @@
+            anims.add(SquooshAnimatedFadeIn(toChild))
+        }
+        prevChild = toChild
+        toChild = toChild.nextSibling
+    }
+
+    // If there are any children remaining in `from` then we need to add them to the start of `to`
+    // and create fade out animations for them.
+    val fromFirstChild = from.firstChild
+    var fromLastChild = from.firstChild
+    while (fromLastChild?.nextSibling != null) {
+        fromLastChild = fromLastChild.nextSibling
+    }
+    // Add fade-outs for all of them.
+    var child = from.firstChild
+    while (child != null) {
+        anims.add(SquooshAnimatedFadeOut(child))
+        child = child.nextSibling
+    }
+    // Move the from remainders into the start of the to list.
+    if (fromLastChild != null) {
+        fromLastChild.nextSibling = to.firstChild
+        to.firstChild = fromFirstChild
+    }
+    from.firstChild = null
+}
+
+private fun extractChildNamed(parent: SquooshResolvedNode, name: String): SquooshResolvedNode? {
+    var child = parent.firstChild
+    var previousChild: SquooshResolvedNode? = null
+    while (child != null) {
+        if (child.view.name == name) {
+            // Remove child from the list.
+            if (previousChild != null) {
+                previousChild.nextSibling = child.nextSibling
+            } else {
+                parent.firstChild = child.nextSibling
+            }
+            child.nextSibling = null
+            return child
+        }
+        previousChild = child
+        child = child.nextSibling
+    }
+    return null
+}
+
+// We can tween between two views (and thus don't need one of them in the tree) if:
+//  - They are both Containers
+//  - They both have a Rect or RoundRect shape (for now we need the shapes to be the same).
+private fun isTweenable(a: View, b: View): Boolean {
+    val aData = a.data
+    val bData = b.data
+    if (aData is ViewData.Container && bData is ViewData.Container) {
+        // Rects and RoundRects can be tweened.
+        if (
+            (aData.shape is ViewShape.Rect && bData.shape is ViewShape.Rect) ||
+                (aData.shape is ViewShape.RoundRect && bData.shape is ViewShape.RoundRect)
+        )
+            return true
+
+        if ((aData.shape is ViewShape.VectorRect && bData.shape is ViewShape.VectorRect))
+            return true
+
+        // Arcs can be tweened.
+        if (aData.shape is ViewShape.Arc && bData.shape is ViewShape.Arc) return true
+    }
+    return false
+}
+
+private fun needsStyleTween(a: ViewStyle, b: ViewStyle): Boolean {
+    // Compare some style things and decide if we need to tween the styles.
+    if (a.background != b.background) return true
+    if (a.stroke != b.stroke) return true
+    return false
+}
+
+// XXX: Horrible code to deal with our terrible generated types. Maybe if style moves to proto then
+//      we'll get some more egonomic generated classes.
+internal fun ViewStyle.asBuilder(): ViewStyle.Builder {
```

Comment:
@rylin8 I think we have another version of this function somewhere else now, too? I should probably delete this version...

Reply:
Yes I copied yours in my PR which is not merged yet -- https://github.com/google/automotive-design-compose/pull/826

Reply:
Your PR has merged, but I haven't updated to use that implementation yet.